Middleware for determining the real ip of the client

MadsRC opened the pull request:
<!--
    Keep PR title verbose enough.
-->

## Changes

This PR adds a middleware which will determine the real ip / original IP of the client based on header values and inject this into the context of the request. A function is also introduced to allow retrieval of this value from the context at a later stage.

When a server is behind a load balancer it is often useful to attempt to determine the IP of the original client so as to not only log the IP of the load balancer. Solutions such as NGINX, AWS ELB and Google Cloud Load Balancers all provide this by appending the IP of the client to an `x-forwarded-for` header. Cloudflare provides this by setting the `True-Client-IP` header to the value of the client.

Only gRPC servers being served over TCP/IP networks are supported.

Implementation of this functionality carries with it the risk of trusting unverified user input and as such allowing the spoofing of source IP's or Denial of Service attacks. Because of this I believe this project should provide a verified and trusted implementation. Such an implementation is included in this PR.

I opted to use a table-driven approach to testing this - I'm unsure if this is acceptable for this repo. If not, please let me know and I'll refactor it.

## Verification

Unit testing with 100% coverage has been implemented. Additionally, I believe I've covered the majority of the edge-cases in testing.

From a security standpoint, the tests verifies that headers are only referenced if the request is coming from a trusted peer. In the event that a trusted peer sends a malformed header, the header is rejected (which is also covered by the tests).

johanbrandhorst commented on interceptors/realip/examples_test.go:
```diff
@@ -0,0 +1,43 @@
+		{IP: net.IPv4(127, 0, 0, 1), Mask: net.IPv4Mask(255, 255, 255, 255)},
+	}
+	// Define headers to look for in the incoming request.
+	headers := []string{"x-forwarded-for", "x-real-ip"}
```
What do you think about defining these header name as constants in the package, to make it easier?

MadsRC replied:
I've added them as package headers now. I originally didn't do it as to not let the package dictate what headers people should use. But I suppose making the well-known defaults available wouldn't hurt.

johanbrandhorst commented on interceptors/realip/realip.go:
```diff
@@ -0,0 +1,116 @@
+	return s.ctx
+}
+
+// UnaryServerInterceptor returns a new unary server interceptors that performs request rate limiting.
```
I think this doc string will need updating 😁

MadsRC replied:
Great catch!

johanbrandhorst commented on interceptors/realip/realip.go:
```diff
@@ -0,0 +1,116 @@
+	}
+}
+
+// StreamServerInterceptor returns a new stream server interceptor that performs rate limiting on the request.
```
Ditto

johanbrandhorst commented on interceptors/realip/realip.go:
```diff
@@ -0,0 +1,116 @@
+
+type realipKey struct{}
+
+func FromContext(ctx context.Context) (net.IP, bool) {
```
What do you think about supporting https://pkg.go.dev/net/netip#Addr? I know `peer.FromContext` returns a `net.IP` but it's sort of deprecated in favor of `netip.Addr`. It should be possible to use https://pkg.go.dev/net/netip#ParseAddr to parse from net.IP.

MadsRC replied:
I've refactored to use netip.Addr ;)

johanbrandhorst reviewed (commented):
This is an _excellent_ PR, thank you! Just a few minor suggestions, let me know what you think.

MadsRC commented:
> This is an _excellent_ PR, thank you! Just a few minor suggestions, let me know what you think.

Thank you! I've implemented the suggestions you had. Turns out the `netip.Addr` rewrite was a bit more involved, but in the end it turned out great. I also rewrote the structure of the tests.

johanbrandhorst reviewed (changes requested):
This looks great, just some more minor points.

MadsRC commented on interceptors/realip/realip.go:
```diff
@@ -0,0 +1,135 @@
+const (
+	X_REAL_IP       = "x-real-ip"
+	X_FORWARDED_FOR = "x-forwarded-for"
+	TRUE_CLIENT_IP  = "true-client-ip"
```
Absolutely - oversight on my part.

johanbrandhorst replied:
Sorry to bang on, but I don't see that the case of the headers was canonicalized?

MadsRC replied:
Hey @johanbrandhorst - I guess I must have misunderstood the meaning of canonicalising the header keys. Let me go ahead and try to fix them ;)

MadsRC replied:
I've changed the casing of the header keys that are the values of the constants - Is that what you meant by canonicalising the header keys?

johanbrandhorst replied:
Yes, that is what I meant. I see you had to lowercase the headers in the map, which is fine, you could also canonicalize them using https://pkg.go.dev/net/textproto#CanonicalMIMEHeaderKey, but this will do.

johanbrandhorst commented on interceptors/realip/realip.go:
```diff
@@ -0,0 +1,135 @@
+	if pr == nil {
+		return noIP
+	}
+	ip, err := netip.ParseAddr(strings.Split(pr.String(), ":")[0])
+	if err != nil {
+		return noIP
+	}
```
Ah, my bad, thanks!

johanbrandhorst reviewed (approved):
LGTM

johanbrandhorst commented:
Thanks for your contribution!
